### ISSUE TITLE
Assertion failed: (castIsValid(op, S, Ty) && "Invalid cast!")

### ISSUE THREAD

**Issue by trasz** [CTSRD-CHERI/llvm#201]
Building CHERIBSD (buildworld) using LLVM with assertions enabled fails like this:

```
Assertion failed: (castIsValid(op, S, Ty) && "Invalid cast!"), function Create, file /home/en322/cheri/llvm/lib/IR/Instructions.cpp, line 2615.
3 warnings generated.
clang-5.0: error: unable to execute command: Abort trap (core dumped)
clang-5.0: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 5.0.0 (https://github.com/CTSRD-CHERI/clang.git a5538873ce268d7f33e6b4343ff0590ba720bb52) (https://github.com/CTSRD-CHERI/llvm.git 926e17c1b5d5110aef32be6b0c7c98054a08c448)

```

https://gist.github.com/trasz/ac2e10f75ec619d08ce300c66997cb4d

**Comment by davidchisnall** [comment]
`-emit-llvm -O0` crashes, so this is a clang bug.

**Comment by davidchisnall** [comment]
Crash is in the call to `kvprintf` in `vsnprintf`.  It appears that the second argument to `kvprintf` is an address-space 0 function pointer.

**Comment by davidchisnall** [comment]
Reduced test case:

```C
void f(void (*func)(void)) [}
```

Produces:

```IR
define void @f(void ()* %func) #0 { ...
```

It appears that this is not constrained to parameter types.  Simpler test case:

```C
void (*func)(void);
```

This generates an AS0 function pointer in a (AS200) global in the IR.

**Comment by kgudka** [comment]
Fixed in CTSRD-CHERI/clang@7783776727ed07a0b4584f781fc7c9040a3d27ac
